src/aggregation/mod.rs

```diff
@@ -0,0 +1,282 @@
+//! This module contains code for aggregation.
+//!
+//! agg_req contains the aggregation request.
+//! agg_req_with_accessor contains the aggregation request plus fast field accessors etc, which are
+//! used during collection.
+//!
+//! segment_agg_result is the aggregation result tree during collection.
+//! intermediate_agg_result is the aggregation tree for merging with other trees.
+//! agg_result is the final aggregation tree.
+
+mod agg_req;
+mod agg_req_with_accessor;
+mod agg_result;
+mod bucket;
+mod executor;
+mod intermediate_agg_result;
+mod metric;
+mod segment_agg_result;
+
+use std::collections::HashMap;
+
+pub use agg_req::Aggregation;
+pub use agg_req::BucketAggregationType;
+pub use agg_req::MetricAggregation;
+use itertools::Itertools;
+
+use self::agg_result::AggregationResults;
+use self::intermediate_agg_result::IntermediateAggregationResults;
+
+/// VecWithNames will be used for th
+#[derive(Clone, PartialEq)]
+pub struct VecWithNames<T: Clone> {
+    data: Vec<T>,
+    data_names: Vec<String>,
+}
+impl<T: Clone> Default for VecWithNames<T> {
+    fn default() -> Self {
+        Self {
+            data: Default::default(),
+            data_names: Default::default(),
+        }
+    }
+}
+
+impl<T: Clone + std::fmt::Debug> std::fmt::Debug for VecWithNames<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("VecWithNames")
+            .field("data", &self.data)
+            .field("data_names", &self.data_names)
+            .finish()
+    }
+}
+
+impl<T: Clone> From<HashMap<String, T>> for VecWithNames<T> {
+    fn from(map: HashMap<String, T>) -> Self {
+        VecWithNames::from_iter(map.into_iter())
+    }
+}
+
+impl<T: Clone> VecWithNames<T> {
+    fn from_iter(iter: impl Iterator<Item = (String, T)>) -> Self {
+        let mut entries = iter.collect_vec();
+        // Sort to ensure order of elements match across multiple instances
+        entries.sort_by_cached_key(|entry| entry.0.to_string());
+        let mut data = vec![];
+        let mut data_names = vec![];
+        for entry in entries {
+            data_names.push(entry.0);
+            data.push(entry.1);
+        }
+        VecWithNames { data, data_names }
+    }
+
+    fn iter_mut(&mut self) -> impl Iterator<Item = (&String, &mut T)> + '_ {
+        self.data_names.iter().zip(self.data.iter_mut())
+    }
+    fn into_iter(self) -> impl Iterator<Item = (String, T)> {
+        self.data_names.into_iter().zip(self.data.into_iter())
+    }
+    fn iter(&self) -> impl Iterator<Item = (&String, &T)> + '_ {
+        self.data_names.iter().zip(self.data.iter())
+    }
+    fn keys(&self) -> impl Iterator<Item = &String> + '_ {
+        self.data_names.iter()
```

Comment:
super nitpick: I think I would still use `Item=&str` here but I agree this is one of the case where it can be argued otherwise.